connectivity/FEATURE_BLE/include/ble/gatt/ChainableGattServerEventHandler.h

```diff
@@ -0,0 +1,211 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2020 ARM Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_CHAINABLEGATTSERVEREVENTHANDLER_H_
+#define MBED_CHAINABLEGATTSERVEREVENTHANDLER_H_
+
+#include "ble/GattServer.h"
+
+/**
+ * GattServer::EventHandler implementation that allows the application
+ * to register multiple separate EventHandlers to be called when
+ * GattServer events happen.
+ */
+class ChainableGattServerEventHandler : public ble::GattServer::EventHandler
+{
+
+public:
+
+    ChainableGattServerEventHandler() { }
+
+    virtual ~ChainableGattServerEventHandler() {
+        // Clean up all nodes
+        auto it = head;
+        while(it) {
+            node_t* temp = it;
+            it = it->next;
+            delete temp;
+        }
+    }
+
+    /**
+     * Register an EventHandler to be notified of events sent to
+     * this ChainableGattServerEventHandler
+     *
+     * @param[in] event_handler Handler to add to chain
+     *
+     * @retval pointer to node on success, nullptr otherwise
+     *
+     * @note the returned node pointer should be retained if
+     * the application ever wishes to deregister the specific EventHandler
+     */
+    void* registerEventHandler(ble::GattServer::EventHandler* event_handler) {
+        auto eh = new (std::nothrow) node_t { event_handler, nullptr };
+        if(!eh) { return nullptr; }
+        if(!head) {
+            head = eh;
+        } else {
+            auto *it = head;
+            while(it->next) {
+                it = it->next;
+            }
+            it->next = eh;
+        }
+        return eh;
+    }
+
+    /**
+     * Deregister an EventHandler previously registered with
+     * registerEventHandler. The return value of that function should
+     * be used to specify which node to remove from the EventHandler chain.
+     *
+     * @param[in] node Pointer to node to deregister
+     */
+    void deregisterEventHandler(void* node) {
```

Comment:
you have to store an array of nodes you registered yourself to be able to remove them, don't think that's necessary the event handler is already a pointer, we could just pass that to remove it

Reply:
True, I have changed this so you just pass in the original event handler pointer.